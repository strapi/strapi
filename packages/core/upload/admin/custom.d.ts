export {};

declare global {
  interface Window {
    strapi: {
      backendURL: string;
    };
  }
  declare module '*?raw';
}
