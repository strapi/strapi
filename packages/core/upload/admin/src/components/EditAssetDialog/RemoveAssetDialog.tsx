import { ConfirmDialog } from '@strapi/admin/strapi-admin';
import { Dialog } from '@strapi/design-system';

import { useRemoveAsset } from '../../hooks/useRemoveAsset';

import type { File } from '../../../../shared/contracts/files';

interface RemoveAssetDialogProps {
  open: boolean;
  onClose: (open: boolean | null) => void;
  asset: File;
}

export const RemoveAssetDialog = ({ open, onClose, asset }: RemoveAssetDialogProps) => {
  // `null` means asset is deleted
  const { removeAsset } = useRemoveAsset(() => {
    onClose(null);
  });

  const handleConfirm = async (event?: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    event?.preventDefault();
    await removeAsset(asset.id);
  };

  return (
    <Dialog.Root open={open} onOpenChange={onClose}>
      <ConfirmDialog onConfirm={handleConfirm} />
    </Dialog.Root>
  );
};
