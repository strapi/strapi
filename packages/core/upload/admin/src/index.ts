import { Images } from '@strapi/icons';

import pluginPkg from '../../package.json';

import { MediaLibraryDialog } from './components/MediaLibraryDialog/MediaLibraryDialog';
import { MediaLibraryInput } from './components/MediaLibraryInput/MediaLibraryInput';
import { PERMISSIONS } from './constants';
import { pluginId } from './pluginId';
import { getTrad, prefixPluginTranslations } from './utils';

import type { MediaLibraryDialogProps } from './components/MediaLibraryDialog/MediaLibraryDialog';
import type { MediaLibraryInputProps } from './components/MediaLibraryInput/MediaLibraryInput';
import type { StrapiApp } from '@strapi/admin/strapi-admin';
import type { Plugin } from '@strapi/types';

const name = pluginPkg.strapi.name;

const admin: Plugin.Config.AdminInput = {
  register(app: StrapiApp) {
    app.addMenuLink({
      to: `plugins/${pluginId}`,
      icon: Images,
      intlLabel: {
        id: `${pluginId}.plugin.name`,
        defaultMessage: 'Media Library',
      },
      permissions: PERMISSIONS.main,
      Component: () => import('./pages/App/App').then((mod) => ({ default: mod.Upload })),
      position: 4,
    });

    app.addSettingsLink('global', {
      id: 'media-library-settings',
      to: 'media-library',
      intlLabel: {
        id: getTrad('plugin.name'),
        defaultMessage: 'Media Library',
      },
      async Component() {
        const { ProtectedSettingsPage } = await import('./pages/SettingsPage/SettingsPage');
        return { default: ProtectedSettingsPage };
      },
      permissions: PERMISSIONS.settings,
    });

    app.addFields({
      type: 'media',
      Component: MediaLibraryInput as React.FC<Partial<MediaLibraryInputProps>>,
    });
    app.addComponents([
      {
        name: 'media-library',
        Component: MediaLibraryDialog as React.FC<Partial<MediaLibraryDialogProps>>,
      },
    ]);

    app.registerPlugin({
      id: pluginId,
      name,
    });
  },
  async registerTrads({ locales }: { locales: string[] }) {
    const importedTrads = await Promise.all(
      locales.map((locale) => {
        return import(`./translations/${locale}.json`)
          .then(({ default: data }) => {
            return {
              data: prefixPluginTranslations(data, pluginId),
              locale,
            };
          })
          .catch(() => {
            return {
              data: {},
              locale,
            };
          });
      })
    );

    return Promise.resolve(importedTrads);
  },
};

// eslint-disable-next-line import/no-default-export
export default admin;
