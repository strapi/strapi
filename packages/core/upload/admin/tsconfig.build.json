{
  "extends": "tsconfig/client.json",
  "compilerOptions": {
    "rootDir": "../",
    "baseUrl": ".",
    "outDir": "../dist",
    "noEmit": false
  },
  "include": ["./src", "./custom.d.ts", "../shared", "../package.json"],
  "exclude": ["**/__mocks__", "./src/**/tests", "**/*.test.*"]
}
