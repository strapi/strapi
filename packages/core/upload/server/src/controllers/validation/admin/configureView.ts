import { yup, validateYupSchema } from '@strapi/utils';
import { ALLOWED_SORT_STRINGS } from '../../../constants';

const configSchema = yup.object({
  pageSize: yup.number().required(),
  sort: yup.mixed().oneOf(ALLOWED_SORT_STRINGS),
});

export const validateViewConfiguration = validateYupSchema(configSchema);

export type ViewConfiguration = yup.InferType<typeof configSchema>;
