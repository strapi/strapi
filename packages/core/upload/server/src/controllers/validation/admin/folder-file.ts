import { intersection, map, isEmpty } from 'lodash/fp';
import { yup, validateYupSchema } from '@strapi/utils';
import { FOLDER_MODEL_UID } from '../../../constants';
import { folderExists } from './utils';
import { isFolderOrChild } from '../../utils/folders';

import type { Folder } from '../../../types';

const validateDeleteManyFoldersFilesSchema = yup
  .object()
  .shape({
    fileIds: yup.array().of(yup.strapiID().required()),
    folderIds: yup.array().of(yup.strapiID().required()),
  })
  .noUnknown()
  .required();

const validateStructureMoveManyFoldersFilesSchema = yup
  .object()
  .shape({
    destinationFolderId: yup
      .strapiID()
      .nullable()
      .defined()
      .test('folder-exists', 'destination folder does not exist', folderExists),
    fileIds: yup.array().of(yup.strapiID().required()),
    folderIds: yup.array().of(yup.strapiID().required()),
  })
  .noUnknown()
  .required();

const validateDuplicatesMoveManyFoldersFilesSchema = yup
  .object()
  .test('are-folders-unique', 'some folders already exist', async function areFoldersUnique(value) {
    const { folderIds, destinationFolderId } = value;
    if (isEmpty(folderIds)) return true;

    const folders = await strapi.db.query(FOLDER_MODEL_UID).findMany({
      select: ['name'],
      where: { id: { $in: folderIds } },
    });

    const existingFolders = await strapi.db.query(FOLDER_MODEL_UID).findMany({
      select: ['name'],
      where: { parent: { id: destinationFolderId } },
    });

    const duplicatedNames = intersection(map('name', folders), map('name', existingFolders));
    if (duplicatedNames.length > 0) {
      return this.createError({
        message: `some folders already exists: ${duplicatedNames.join(', ')}`,
      });
    }

    return true;
  });

const validateMoveFoldersNotInsideThemselvesSchema = yup
  .object()
  .test(
    'dont-move-inside-self',
    'folders cannot be moved inside themselves or one of its children',
    async function validateMoveFoldersNotInsideThemselves(value) {
      const { folderIds, destinationFolderId } = value;
      if (destinationFolderId === null || isEmpty(folderIds)) return true;

      const destinationFolder = await strapi.db.query(FOLDER_MODEL_UID).findOne({
        select: ['path'],
        where: { id: destinationFolderId },
      });

      const folders: Folder[] = await strapi.db.query(FOLDER_MODEL_UID).findMany({
        select: ['name', 'path'],
        where: { id: { $in: folderIds } },
      });

      const unmovableFoldersNames = folders
        .filter((folder) => isFolderOrChild(destinationFolder, folder))
        .map((f) => f.name);

      if (unmovableFoldersNames.length > 0) {
        return this.createError({
          message: `folders cannot be moved inside themselves or one of its children: ${unmovableFoldersNames.join(
            ', '
          )}`,
        });
      }

      return true;
    }
  );

export const validateDeleteManyFoldersFiles = validateYupSchema(
  validateDeleteManyFoldersFilesSchema
);

export async function validateMoveManyFoldersFiles(body: unknown) {
  await validateYupSchema(validateStructureMoveManyFoldersFilesSchema)(body);
  await validateYupSchema(validateDuplicatesMoveManyFoldersFilesSchema)(body);
  await validateYupSchema(validateMoveFoldersNotInsideThemselvesSchema)(body);
}
