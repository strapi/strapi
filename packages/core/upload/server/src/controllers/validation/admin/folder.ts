import { isUndefined, get, isNil } from 'lodash/fp';
import { yup, validateYupSchema } from '@strapi/utils';
import { getService } from '../../../utils';
import { FOLDER_MODEL_UID } from '../../../constants';
import { folderExists } from './utils';
import { isFolderOrChild } from '../../utils/folders';

const NO_SLASH_REGEX = /^[^/]+$/;
const NO_SPACES_AROUND = /^(?! ).+(?<! )$/;

const isNameUniqueInFolder = (id?: number): yup.TestFunction<string | undefined> => {
  return async function test(name?: string) {
    const { exists } = getService('folder');
    const filters: Record<string, unknown> = { name, parent: this.parent.parent || null };
    if (id) {
      filters.id = { $ne: id };

      if (isUndefined(name)) {
        const existingFolder = await strapi.db.query(FOLDER_MODEL_UID).findOne({ where: { id } });
        filters.name = get('name', existingFolder);
      }
    }

    const doesExist = await exists(filters);
    return !doesExist;
  };
};

const validateCreateFolderSchema = yup
  .object()
  .shape({
    name: yup
      .string()
      .min(1)
      .matches(NO_SLASH_REGEX, 'name cannot contain slashes')
      .matches(NO_SPACES_AROUND, 'name cannot start or end with a whitespace')
      .required()
      .test('is-folder-unique', 'A folder with this name already exists', isNameUniqueInFolder()),
    parent: yup
      .strapiID()
      .nullable()
      .test('folder-exists', 'parent folder does not exist', folderExists),
  })
  .noUnknown()
  .required();

const validateUpdateFolderSchema = (id: number) =>
  yup
    .object()
    .shape({
      name: yup
        .string()
        .min(1)
        .matches(NO_SLASH_REGEX, 'name cannot contain slashes')
        .matches(NO_SPACES_AROUND, 'name cannot start or end with a whitespace')
        .test(
          'is-folder-unique',
          'A folder with this name already exists',
          isNameUniqueInFolder(id)
        ),
      parent: yup
        .strapiID()
        .nullable()
        .test('folder-exists', 'parent folder does not exist', folderExists)
        .test(
          'dont-move-inside-self',
          'folder cannot be moved inside itself',
          async function test(parent) {
            if (isNil(parent)) return true;

            const destinationFolder = await strapi.db.query(FOLDER_MODEL_UID).findOne({
              select: ['path'],
              where: { id: parent },
            });

            const currentFolder = await strapi.db.query(FOLDER_MODEL_UID).findOne({
              select: ['path'],
              where: { id },
            });

            if (!destinationFolder || !currentFolder) return true;

            return !isFolderOrChild(destinationFolder, currentFolder);
          }
        ),
    })
    .noUnknown()
    .required();

export const validateCreateFolder = validateYupSchema(validateCreateFolderSchema);
export const validateUpdateFolder = (id: number) =>
  validateYupSchema(validateUpdateFolderSchema(id));
