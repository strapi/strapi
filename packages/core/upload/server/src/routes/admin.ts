export const routes = {
  type: 'admin',
  routes: [
    {
      method: 'GET',
      path: '/settings',
      handler: 'admin-settings.getSettings',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.settings.read'],
            },
          },
        ],
      },
    },
    {
      method: 'PUT',
      path: '/settings',
      handler: 'admin-settings.updateSettings',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.settings.read'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/',
      handler: 'admin-upload.upload',
      config: {
        policies: ['admin::isAuthenticatedAdmin'],
      },
    },
    {
      method: 'GET',
      path: '/files',
      handler: 'admin-file.find',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/files/:id',
      handler: 'admin-file.findOne',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'DELETE',
      path: '/files/:id',
      handler: 'admin-file.destroy',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/folders/:id',
      handler: 'admin-folder.findOne',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/folders',
      handler: 'admin-folder.find',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/folders',
      handler: 'admin-folder.create',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.create'],
            },
          },
        ],
      },
    },
    {
      method: 'PUT',
      path: '/folders/:id',
      handler: 'admin-folder.update',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/folder-structure',
      handler: 'admin-folder.getStructure',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/actions/bulk-delete',
      handler: 'admin-folder-file.deleteMany',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/actions/bulk-move',
      handler: 'admin-folder-file.moveMany',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
  ],
};
