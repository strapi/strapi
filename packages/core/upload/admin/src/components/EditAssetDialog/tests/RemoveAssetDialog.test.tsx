import { configureStore } from '@reduxjs/toolkit';
import { NotificationsProvider, adminApi } from '@strapi/admin/strapi-admin';
import { DesignSystemProvider } from '@strapi/design-system';
import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { IntlProvider } from 'react-intl';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';

import en from '../../../translations/en.json';
import { RemoveAssetDialog } from '../RemoveAssetDialog';

type Messages = typeof en;
type MessageKeys = keyof Messages;

const enKeys = Object.keys(en) as MessageKeys[];

const messageForPlugin = enKeys.reduce(
  (acc: { [key in MessageKeys]: string }, curr: MessageKeys) => {
    acc[curr] = `upload.${en[curr]}`;

    return acc;
  },
  {} as { [key in MessageKeys]: string }
);

const asset = {
  id: 8,
  name: 'Screenshot 2.png',
  alternativeText: null,
  caption: null,
  width: 1476,
  height: 780,
  formats: {
    thumbnail: {
      name: 'thumbnail_Screenshot 2.png',
      hash: 'thumbnail_Screenshot_2_5d4a574d61',
      ext: '.png',
      mime: 'image/png',
      width: 245,
      height: 129,
      size: 10.7,
      path: null,
      url: '/uploads/thumbnail_Screenshot_2_5d4a574d61.png',
    },
    large: {
      name: 'large_Screenshot 2.png',
      hash: 'large_Screenshot_2_5d4a574d61',
      ext: '.png',
      mime: 'image/png',
      width: 1000,
      height: 528,
      size: 97.1,
      path: null,
      url: '/uploads/large_Screenshot_2_5d4a574d61.png',
    },
    medium: {
      name: 'medium_Screenshot 2.png',
      hash: 'medium_Screenshot_2_5d4a574d61',
      ext: '.png',
      mime: 'image/png',
      width: 750,
      height: 396,
      size: 58.7,
      path: null,
      url: '/uploads/medium_Screenshot_2_5d4a574d61.png',
    },
    small: {
      name: 'small_Screenshot 2.png',
      hash: 'small_Screenshot_2_5d4a574d61',
      ext: '.png',
      mime: 'image/png',
      width: 500,
      height: 264,
      size: 31.06,
      path: null,
      url: '/uploads/small_Screenshot_2_5d4a574d61.png',
    },
  },
  hash: 'Screenshot_2_5d4a574d61',
  ext: '.png',
  mime: 'image/png',
  size: 102.01,
  url: '/uploads/Screenshot_2_5d4a574d61.png',
  previewUrl: null,
  provider: 'local',
  provider_metadata: null,
  createdAt: '2021-10-04T09:42:31.670Z',
  updatedAt: '2021-10-04T09:42:31.670Z',
};

const store = configureStore({
  reducer: { [adminApi.reducerPath]: adminApi.reducer },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(adminApi.middleware),
});

const ComponentFixture = ({ children }: { children: React.ReactNode }) => (
  <Provider store={store}>
    <QueryClientProvider
      client={
        new QueryClient({
          defaultOptions: {
            queries: {
              refetchOnWindowFocus: false,
            },
          },
        })
      }
    >
      <DesignSystemProvider>
        <IntlProvider locale="en" messages={messageForPlugin} defaultLocale="en">
          <NotificationsProvider>{children}</NotificationsProvider>
        </IntlProvider>
      </DesignSystemProvider>
    </QueryClientProvider>
  </Provider>
);

const renderCompo = (handleCloseSpy = jest.fn()) => {
  return render(
    <ComponentFixture>
      <RemoveAssetDialog open onClose={handleCloseSpy} asset={asset} />
    </ComponentFixture>
  );
};

const user = userEvent.setup();

describe('RemoveAssetDialog', () => {
  it('snapshots the component', () => {
    renderCompo();

    expect(document.body).toMatchSnapshot();
  });

  it('closes the dialog when pressing cancel', async () => {
    const handleCloseSpy = jest.fn();
    renderCompo(handleCloseSpy);

    await user.click(screen.getByText('Cancel'));

    expect(handleCloseSpy).toHaveBeenCalled();
  });
});
