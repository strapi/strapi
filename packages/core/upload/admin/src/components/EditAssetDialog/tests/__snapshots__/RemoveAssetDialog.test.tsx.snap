// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveAssetDialog snapshots the component 1`] = `
.c0 {
  margin-inline-start: -250px;
  top: 4.6rem;
  left: 50%;
  width: 50rem;
  z-index: 700;
  position: fixed;
}

.c5 {
  padding-block-start: 24px;
  padding-inline-end: 24px;
  padding-block-end: 24px;
  padding-inline-start: 24px;
}

.c8 {
  padding-block-start: 40px;
  padding-block-end: 40px;
  padding-inline-start: 24px;
  padding-inline-end: 24px;
}

.c12 {
  padding-block-start: 16px;
  padding-inline-end: 16px;
  padding-block-end: 16px;
  padding-inline-start: 16px;
}

.c15 {
  padding-inline-start: 16px;
  padding-inline-end: 16px;
  width: 100%;
  border-radius: 4px;
  cursor: pointer;
}

.c1 {
  gap: 8px;
  align-items: stretch;
  flex-direction: column;
  display: flex;
}

.c9 {
  gap: 8px;
  align-items: center;
  flex-direction: column;
  display: flex;
}

.c13 {
  gap: 8px;
  align-items: center;
  justify-content: space-between;
  flex-direction: row;
  display: flex;
}

.c16 {
  gap: 8px;
  align-items: center;
  justify-content: center;
  flex-direction: row;
  display: inline-flex;
}

.c6 {
  font-weight: 600;
  font-size: 1.8rem;
  line-height: 1.22;
  color: currentcolor;
  font-weight: 600;
}

.c11 {
  font-size: 1.4rem;
  line-height: 1.43;
  color: currentcolor;
}

.c18 {
  font-size: 1.2rem;
  line-height: 1.33;
  color: currentcolor;
  font-weight: 600;
}

.c2 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c17 {
  height: 3.2rem;
  text-decoration: none;
  border: 1px solid #dcdce4;
  background: #ffffff;
  color: #32324d;
}

.c17:hover {
  background-color: #f6f6f9;
}

.c17:active {
  background-color: #eaeaef;
}

.c17[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c19 {
  height: 3.2rem;
  text-decoration: none;
  border: 1px solid #f5c0b8;
  background: #fcecea;
  color: #b72b1a;
}

.c19:hover {
  background-color: #ffffff;
}

.c19:active {
  background-color: #ffffff;
  border: 1px solid #d02b20;
  color: #d02b20;
}

.c19[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c3 {
  background: #32324d33;
  position: fixed;
  inset: 0;
  z-index: 300;
  will-change: opacity;
}

.c4 {
  max-width: 42rem;
  height: min-content;
  width: 100%;
  overflow: hidden;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  border-radius: 4px;
  background-color: #ffffff;
  box-shadow: 0px 2px 15px rgba(33, 33, 52, 0.1);
  z-index: 310;
}

.c7 {
  display: flex;
  justify-content: center;
  border-bottom: solid 1px #eaeaef;
}

.c14 {
  border-top: solid 1px #eaeaef;
  flex: 1;
}

.c10 {
  width: 24px;
  height: 24px;
}

.c10 path {
  fill: #d02b20;
}

@media (prefers-reduced-motion: no-preference) {
  .c17 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c19 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c3 {
    animation: ececrQ 200ms cubic-bezier(.4,0,.2,1);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c4[data-state='open'] {
    animation-duration: 200ms;
    animation-timing-function: cubic-bezier(.4,0,.2,1);
    animation-name: iWCnzW;
  }

  .c4[data-state='closed'] {
    animation-duration: 120ms;
    animation-timing-function: cubic-bezier(0.25, 0.46, 0.45, 0.94);
    animation-name: jgufbq;
  }
}

<body
  data-scroll-locked="1"
  style="pointer-events: none;"
>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div>
    <div
      aria-hidden="true"
      class="c0 c1"
      data-aria-hidden="true"
    />
    <span
      class="c2"
    >
      <p
        aria-live="polite"
        aria-relevant="all"
        id="live-region-log"
        role="log"
      />
      <p
        aria-live="polite"
        aria-relevant="all"
        id="live-region-status"
        role="status"
      />
      <p
        aria-live="assertive"
        aria-relevant="all"
        id="live-region-alert"
        role="alert"
      />
    </span>
  </div>
  <div
    class="c3"
    data-state="open"
    style="pointer-events: auto;"
  >
    <div
      aria-describedby="radix-:r2:"
      aria-labelledby="radix-:r1:"
      class="c4"
      data-state="open"
      id="radix-:r0:"
      role="alertdialog"
      style="pointer-events: auto;"
      tabindex="-1"
    >
      <h2
        class="c5 c6 c7"
        id="radix-:r1:"
      >
        Confirmation
      </h2>
      <div
        class="c8 c9"
      >
        <svg
          class="c10"
          fill="currentColor"
          height="24"
          viewBox="0 0 32 32"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 3a13 13 0 1 0 13 13A13.013 13.013 0 0 0 16 3m-1 7a1 1 0 0 1 2 0v7a1 1 0 0 1-2 0zm1 13a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3"
          />
        </svg>
        <p
          class="c11"
          id="radix-:r2:"
        >
          Are you sure?
        </p>
      </div>
      <footer
        class="c12 c13 c14"
      >
        <button
          aria-disabled="false"
          class="c15 c16 c17"
          type="button"
        >
          <span
            class="c18"
          >
            Cancel
          </span>
        </button>
        <button
          aria-disabled="false"
          class="c15 c16 c19"
          type="button"
        >
          <span
            class="c18"
          >
            Confirm
          </span>
        </button>
      </footer>
    </div>
  </div>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;
